engine/consensus/dkg/dkg_controller_test.go

```diff
@@ -0,0 +1,273 @@
+package dkg
+
+import (
+	"crypto/rand"
+	"fmt"
+	"os"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/rs/zerolog"
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/flow-go/crypto"
+)
+
+// node is a test object that simulates a running instance of the DKG protocol
+// where transitions from one phase to another are dictated by a timer.
+type node struct {
+	id             int
+	controller     *Controller
+	phase0Duration time.Duration
+	phase1Duration time.Duration
+	phase2Duration time.Duration
+}
+
+func newNode(id int, controller *Controller,
+	phase0Duration time.Duration,
+	phase1Duration time.Duration,
+	phase2Duration time.Duration) *node {
+
+	return &node{
+		id:             id,
+		controller:     controller,
+		phase0Duration: phase0Duration,
+		phase1Duration: phase1Duration,
+		phase2Duration: phase2Duration,
+	}
+}
+
+func (n *node) run() error {
+	// start the DKG controller
+	go func() {
+		_ = n.controller.Run()
+	}()
+
+	// Wait and trigger transition from Phase 0 to Phase 1
+	err := n.delayedTask(n.phase0Duration, n.controller.EndPhase0)
+	if err != nil {
+		return err
+	}
+
+	// Wait and trigger transition from Phase 1 to Phase 2
+	err = n.delayedTask(n.phase1Duration, n.controller.EndPhase1)
+	if err != nil {
+		return err
+	}
+
+	// Wait and retrieve DKG results
+	err = n.delayedTask(
+		n.phase2Duration,
+		func() error {
+			err := n.controller.End()
+			return err
+		})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (n *node) delayedTask(delay time.Duration, task func() error) error {
+	timer := time.After(delay)
+	<-timer
+	return task()
+}
+
+// processor is an implementation of DKGProcessor that enables nodes to exchange
+// private and public messages.
+type processor struct {
+	id       int
+	channels []chan DKGMessage
+}
+
+func (proc *processor) PrivateSend(dest int, data []byte) {
+	proc.channels[dest] <- DKGMessage{Orig: proc.id, Data: data}
+}
+
+// ATTENTION: Normally the processor requires Broadcast to provide guaranteed
+// delivery (either all nodes receive the message or none of them receive it).
+// Here we are just assuming that with a long enough duration for phases 1 and
+// 2, all nodes are guaranteed to see everyone's messages. So it is important
+// to set timeouts carefully in the tests.
+func (proc *processor) Broadcast(data []byte) {
+	for i := 0; i < len(proc.channels); i++ {
+		if i == proc.id {
+			continue
+		}
+		proc.channels[i] <- DKGMessage{Orig: proc.id, Data: data}
+	}
+}
+
+func (proc *processor) Blacklist(node int) {}
```

Comment:
It would be good to log here to help debugging failing tests. That's actually what happened when the phase is too short. 
Ex: `(node %d blacklisted node %d, proc.id, node)`. 

I'll also add an argument to this function in the crypto lib to indicate the violation.